staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go

```diff
@@ -536,11 +542,11 @@ type validationActivation struct {
 
 func validationActivationWithOldSelf(sts *schema.Structural, obj, oldObj interface{}) interpreter.Activation {
 	va := &validationActivation{
-		self: UnstructuredToVal(obj, sts),
+		self:       UnstructuredToVal(obj, sts),
+		hasOldSelf: true,
```

Comment:
hasOldSelf always be true even oldObj == nil? Which means the transaction rule will always be turned on. I am not sure if that makes sense for user to be able to use oldSelf without having an existing old object?

Reply:
> I am not sure if that makes sense for user to be able to use oldSelf without having an existing old object?

That is the very intention of this feature. Rules that have `optionalOldSelf` are evaluated with `null` `oldSelf`. If `optionalOldSelf` is false (the default, as it is today for transition rules), then the rule is never evaluated.

Reply:
> set hasOldSelf to be always true might potentially break the existing feature behavior

It is writen so that the existing feature remains the same. If expressions dont mention `oldSelf` then they can never observe `hasOldSelf`. A transition rule will also work the same, since it is not evaluated when `oldSelf` is nil in the first place.